jl4-lsp/src/LSP/L4/Rules.hs

```diff
@@ -741,3 +723,56 @@ rangeOfResolveWarning = \case
     srcSpanToLspRange Nothing
   Resolve.Ambiguous name _ ->
     srcRangeToLspRange $ rangeOf name
+
+-- ----------------------------------------------------------------------------
+-- Helpers for implementing syntax highlighting
+-- ----------------------------------------------------------------------------
+
+applyPositionMapping :: [SemanticToken] -> PositionMapping -> [SemanticToken]
+applyPositionMapping semTokens positionMapping =
+  Maybe.mapMaybe
+    ( \t ->
+        case toCurrentPosition positionMapping t.start of
+          Nothing -> Nothing
+          Just newPos -> Just (t & #start .~ newPos)
+    )
+    semTokens
+
+-- | @'semanticTokensUsing' phases@
+--
+-- Helper function for defining multi-phase semantic syntax highlighting.
+--
+-- Each phase can produce '[SemanticToken]'s and 'PositionMapping' if the result is outdated.
+-- Tokens obtained from earlier phases take precedence over tokens from later phases.
+--
+-- TODO: this function is inefficient, as all phases are always executed.
+-- It would be better if we could tell whether the result is actually stale and then
+-- we won't run later phases.
+-- E.g. @NormalizedUri -> Action (Maybe ([SemanticToken], Maybe PositionMapping))@
+semanticTokensUsing ::
+  [NormalizedUri -> Action (Maybe ([SemanticToken], PositionMapping))] ->
+  (NormalizedUri -> Action [SemanticToken])
+semanticTokensUsing phases uri = do
+  semTokens <- traverse ($ uri) phases
+  let
+    pmSemTokens = fmap (uncurry applyPositionMapping) $ catMaybes semTokens
+
+  case pmSemTokens of
+    [] -> pure []
+    x : xs -> pure $ foldl mergeSameLengthTokens x xs
+ where
+  -- We assume that semantic tokens do *not* change its length, no matter whether they
+  -- have been lexed, parsed or typechecked.
+  -- A rather bold assumption, tbh. It will almost definitely not hold
+  -- up in practice, but let's do one step at a time.
+  mergeSameLengthTokens :: [SemanticToken] -> [SemanticToken] -> [SemanticToken]
+  mergeSameLengthTokens [] bs = bs
+  mergeSameLengthTokens as [] = as
+  mergeSameLengthTokens (a : as) (b : bs) = case compare a.start b.start of
+    -- a.start == b.start
+    -- Same token, only print one
+    EQ -> a : mergeSameLengthTokens as bs
+    -- a.start < b.start
+    LT -> a : mergeSameLengthTokens as (b : bs)
+    -- a.start > b.start
+    GT -> b : mergeSameLengthTokens (a : as) bs
```

Comment:
the part that contains the actual logic is  a pure function. I think that fact should be reflected in the code. 

Tbh I also don't understand why this is correct. Doesn't merging in that way require that the tokens are ordered by length? Is that the case?

Reply:
Oh yeah start is what I meant! Yeah then this makes sense! Maybe you should add a comment „tokens are ordered by the length of the starting symbol“